### ISSUE TITLE
Implementation of enum in migration

### ISSUE THREAD

**Issue by Dublerq** [phalcon/phalcon-devtools#562]
Migration enum columns change the type from enum(str,str2...) to char(1), please fix it.

**Comment by mattvb91** [comment]
Can anyone else help with the regex in zephir to get just the values out of something like:

enum('test', 'test1', 'test(2)', 'test3')

**Comment by sergeyklay** [comment]
@mattvb91 I think for questions like this there are dedicated regexp forums and the Zephir repo

**Comment by mattvb91** [comment]
Thanks @sergeyklay will check in there, thought it would be relevant here as its for the implementation of  this issue :+1:

**Comment by marhub** [comment]
Two months and nothing? C'mon!

**Comment by sergeyklay** [comment]
You can to do it. I'll try to review

**Comment by mattvb91** [comment]
I have a topic open here:
http://forum.zephir-lang.com/discussion/3477/regex-escaping-help

The only issue I need help with is how zephir is escaping the regex, apart from that I could have this committed already :+1: 

In a nutshell I cant do this

     \((.*)\)$

Due to how zephir handles escaped strings. Once someone can explain it to me I can close this.

**Comment by mattvb91** [comment]
https://github.com/phalcon/cphalcon/pull/11368

**Comment by ChangePlaces** [comment]
dupes: 
- #10216
- #11139

**Comment by DroneZzZko** [comment]
Any updates? 2 years have passed...

**Comment by sergeyklay** [comment]
@sergeysviridenko Could you please deal with this

**Comment by sergeysviridenko** [comment]
@DroneZzZko I'll try to sort out with this as soon as possible.

**Comment by markhilton** [comment]
hello guys! any updated on fixing ENUM, maybe ETA?

**Comment by sergeysviridenko** [comment]
@markhilton Sorry, but at the moment there isn't any update. Enum field is spec field for MySQL db. Command migration creates files which can be use on different DB. For example - if we create migration with `ENUM` field, we won't use it on PostgreSQL db. Do you have any proposition about implementing this field?

**Comment by markhilton** [comment]
@sergeysviridenko nope, I just ported a database to Phalcon and noticed this error while doing migrations, so wanted to get an update to determine direction for my project. Thanks for taking time to answer, appreciate it!

**Comment by amnuts** [comment]
@sergeysviridenko I wouldn't say that they're specific to MySQL.  [PostgreSQL has enums](https://www.postgresql.org/docs/8.3/static/datatype-enum.html), for example.  However, beyond that others mayb not (Oracle or SqlServer) for example.

Perhaps you could generate a commented out block for the specific db type on which the migration generation was run?  Chances are someone is doing that because they're wanting to track the changes to their db structure and will retain that db type rather than running migration in order to change db type.  So having the specific items in there - even if commented out - would be really handy for the user who can then just remove the comments.

For example:

```
    public function morph()
    {
        $this->morphTable('users', [
                'columns' => [
                    new Column(
                        'id',
                        [
                            'type' => Column::TYPE_INTEGER,
                            'unsigned' => true,
                            'notNull' => true,
                            'autoIncrement' => true,
                            'size' => 11,
                            'first' => true
                        ]
                    ),
                    /*
                     * MySQL DB migration
                     * This field type was automatically generated from a MySQL
                     * instance.  Using this to migrate to another database type
                     * is not guaranteed to work.  We have supplied a generic
                     * version below, or you can manually replace the generic
                     * field with the following if you remain on MySQL:
                     *
                     *    new Column(
                     *       'level',
                     *       [
                     *           'type' => Column::TYPE_ENUM,
                     *           'default' => "guest",
                     *           'notNull' => true,
                     *           'size' => 1,
                     *           'after' => 'id',
                     *           'value' => "'guest','user','super','admin'"
                     *       ]
                     *   )
                     *
                     */
                    new Column(
                        'level',
                        [
                            'type' => Column::TYPE_CHAR,
                            'default' => "none",
                            'notNull' => true,
                            'size' => 1,
                            'after' => 'id'
                        ]
                    )
                ],
                'indexes' => [
                    new Index('PRIMARY', ['id'], 'PRIMARY')
                ],
                'options' => [
                    'TABLE_TYPE' => 'BASE TABLE',
                    'AUTO_INCREMENT' => '',
                    'ENGINE' => 'InnoDB',
                    'TABLE_COLLATION' => 'utf8mb4_bin'
                ],
            ]
        );
    }
```

From the documentation it looks like you support some MySQL & PostgeSQL specific things such as onDelete/onUpdate, so it wouldn't be too far outside of that to put in column-specific stuff even without the comment.

**Comment by ChangePlaces** [comment]
Wow, I'm surprised this project is still going! I thought it was dead!

**Comment by amnuts** [comment]
@ChangePlaces hope not, because I've only just started to use Phalcon and its migrations!  Does seem to have gone on for quite some time, though - a couple years at least, it seems.  There seems to be an assumption that the tool can only be generic and support all databases and not give the specifics that the user may choose to use because it suits their needs.  I'd rather it works so that I can track all the changes required in the db structure for my project, which runs on the db platform of my choosing, and will never be distributed - which means db-specific support where needed.

**Comment by beatwade** [comment]
@markhilton Hi, you can use query in migration with up function. this is very flexible.
such as this
``php
    /**
     * Define the table structure
     *
     * @return void
     */
    public function morph()
    {
        $this->morphTable('user_house_privilege', [
                'columns' => [
                    new Column(
                        'id',
                        [
                            'type' => Column::TYPE_VARCHAR,
                            'notNull' => true,
                            'size' => 100,
                            'first' => true
                        ]
                    ),
                    new Column(
                        'house_id',
                        [
                            'type' => Column::TYPE_INTEGER,
                            'notNull' => false,
                            'size' => 100,
                            'after' => 'id'
                        ]
                    ),
                    new Column(
                        'user_id',
                        [
                            'type' => Column::TYPE_INTEGER,
                            'notNull' => false,
                            'size' => 100,
                            'after' => 'house_id'
                        ]
                    ),
                    new Column(
                        'privilege_ids',
                        [
                            'type' => Column::TYPE_VARCHAR,
                            'size' => 100,
                            'notNull' => false,
                            'after' => 'user_id'
                        ]
                    )
                ],
                'indexes' => [
                    new Index('PRIMARY', ['id'], 'PRIMARY'),
                    new Index('house_user_index', ['user_id', 'house_id'], null)
                ],
                'options' => [
                    'TABLE_TYPE' => 'BASE TABLE',
                    'AUTO_INCREMENT' => '',
                    'ENGINE' => 'InnoDB',
                    'TABLE_COLLATION' => 'utf8_general_ci'
                ],
            ]
        );
    }

    /**
     * Run the migrations
     *
     * @return void
     */
    public function up()
    {
        self::$_connection->query("ALTER TABLE user_house_privilege ADD COLUMN user_role ENUM('renter','owner')");
    }

    /**
     * Reverse the migrations
     *
     * @return void
     */
    public function down()
    {
        self::$_connection->dropTable("user_house_privilege");
    }

``

**Comment by beatwade** [comment]
@markhilton you can use connection with query in migration up method

**Comment by sergeyklay** [comment]
@sergeysviridenko Could you please deal with this

**Comment by sergeysviridenko** [comment]
Ok. I'll try to sort out.

**Comment by sergeysviridenko** [comment]
We'll try to implement in 4.0.x branch

**Comment by cryptolio** [comment]
Any quick fix for this stupid issue?

**Comment by sergeysviridenko** [comment]
@cryptolio there it can't be quick. We will try to implement it in 4.0.0 branch.

**Comment by maxttor** [comment]
What about SET type?
